src/Microsoft.AspNetCore.ResponseCaching/ResponseOutput/ReadOnlyShardStream.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.ResponseCaching.Internal
+{
+    internal class ReadOnlyShardStream : Stream
+    {
+        private readonly List<byte[]> _shards;
+        private readonly long _length;
+        private readonly int _shardSize;
+        private int _shardIndex;
+        private int _shardOffset;
+        private long _position;
+
+        internal ReadOnlyShardStream(List<byte[]> shards, int shardSize, long length)
```

Comment:
Assuming shardSize and length are not lies. Given that this is internal though, I have not added logic to check these values.